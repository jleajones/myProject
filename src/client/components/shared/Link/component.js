import React from 'react';
import PropTypes from 'prop-types';
import { withRouter, Link as RouterLink } from 'react-router-dom';
import MUILink from '@material-ui/core/Link';

const EVENT_NAME = `Click`;

/**
 *
 * @param category
 * @param children
 * @param color
 * @param dispatchInteraction
 * @param history
 * @param label
 * @param onClick
 * @param to
 * @param uuid
 * @param className
 * @param component
 * @returns {*}
 * @constructor
 */
const Link = ({
  category,
  children,
  color,
  dispatchInteraction,
  history,
  label,
  onClick,
  to,
  uuid,
  className,
  component
}) => {
  const onClickFn = e => {
    e.preventDefault();
    dispatchInteraction({
      eventName: EVENT_NAME,
      label: `${label}-link`,
      category,
      url: window.location.href,
      uuid,
      properties: {}
    });

    if (onClick) onClick(e);
    history.push(to);
  };

  return (
    <RouterLink onClick={onClickFn} to={to}>
      <MUILink color={color} className={className} component={component}>
        {children}
      </MUILink>
    </RouterLink>
  );
};

Link.defaultProps = {
  category: '',
  className: '',
  onClick: () => {},
  color: 'inherit',
  component: 'span'
};

Link.propTypes = {
  label: PropTypes.string.isRequired,
  uuid: PropTypes.string.isRequired,
  className: PropTypes.string,
  children: PropTypes.node.isRequired,
  to: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.shape({
      pathname: PropTypes.string,
      search: PropTypes.string,
      hash: PropTypes.string,
      state: PropTypes.string
    })
  ]).isRequired,
  color: PropTypes.string,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }).isRequired,
  category: PropTypes.string,
  component: PropTypes.string,
  onClick: PropTypes.func,
  dispatchInteraction: PropTypes.func.isRequired
};

export default withRouter(Link);
