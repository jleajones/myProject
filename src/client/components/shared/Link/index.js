import React from 'react';
import PropTypes from 'prop-types';
import { withRouter, Link as OriginalLink } from 'react-router-dom';
import { default as MUILink } from '@material-ui/core/Link';

import { trackEvent } from '@client/lib/tracking';

const EVENT_NAME = `LinkClick`;

const Link = ({ children, label, category, onClick, to, history, color, ...rest }) => {
  const onClickFn = e => {
    e.preventDefault();
    trackEvent(EVENT_NAME, {
      label,
      category
    });

    if (onClick) onClick(e);
    history.push(to);
  };

  return (
    <OriginalLink onClick={onClickFn} to={to}>
      <MUILink color={color} component="span">
        {children}
      </MUILink>
    </OriginalLink>
  );
};

Link.propTypes = {
  label: PropTypes.string.isRequired,
  children: PropTypes.node.isRequired,
  category: PropTypes.string,
  onClick: PropTypes.func
};

export default withRouter(Link);
