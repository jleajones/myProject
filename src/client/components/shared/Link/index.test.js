import React from 'react';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';

import Link from '.';

describe('<Link />', () => {
  let dispatchInteraction;
  let history;
  let label;
  let uuid;
  let category;
  let to;
  let store;
  let initialState;

  beforeEach(() => {
    dispatchInteraction = jest.fn();
    history = {};
    label = 'some label';
    uuid = 'aaaa-sss-asdfd-asdgsd';
    category = 'someCategory';
    to = '/some/url';
    store = configureStore();
    initialState = {
      analytics: {
        uuid
      }
    };
  });
  test('Renders component without crashing', () => {
    const component = shallow(
      <Link
        store={store(initialState)}
        dispatchInteraction={dispatchInteraction}
        history={history}
        label={label}
        category={category}
        to={to}
      >
        Some Text
      </Link>
    );

    expect(component).toMatchSnapshot();
  });
});
