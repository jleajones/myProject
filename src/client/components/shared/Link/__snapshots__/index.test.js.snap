// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Link /> Renders component without crashing 1`] = `
<ContextProvider
  value={null}
>
  <withRouter(Link)
    category="someCategory"
    dispatchInteraction={[Function]}
    history={Object {}}
    label="some label"
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    to="/some/url"
    uuid="aaaa-sss-asdfd-asdgsd"
  >
    Some Text
  </withRouter(Link)>
</ContextProvider>
`;
